locale.go

```diff
@@ -0,0 +1,200 @@
+	allDigits   = regexp.MustCompile(`^[0-9]*$`)
+)
+
+// normalizeCodeset produces a normalized version of a codeset
+// Based on libintl's _nl_normalize_codeset
+func normalizeCodeset(codeset string) string {
+	codeset = strings.ToLower(notAlphaNum.ReplaceAllLiteralString(codeset, ""))
+	if allDigits.MatchString(codeset) {
+		codeset = "iso" + codeset
+	}
+	return "." + codeset
+}
+
+const (
+	xpgNormCodeset = 1 << iota
+	xpgCodeset
+	xpgTerritory
+	xpgModifier
+)
+
+func expandLocale(locale string) (locales []string) {
+	locale = replaceAliases(locale)
+
+	// Split the locale based on libintl's _nl_explode_name()
+	var language, territory, codeset, normCodeset, modifier string
+	mask := 0
+	pos := strings.IndexRune(locale, '@')
+	if pos != -1 {
+		modifier = locale[pos:]
+		locale = locale[:pos]
+		mask |= xpgModifier
+	}
+	pos = strings.IndexRune(locale, '.')
+	if pos != -1 {
+		codeset = locale[pos:]
+		locale = locale[:pos]
+		mask |= xpgCodeset
+
+		normCodeset = normalizeCodeset(codeset)
+		if normCodeset != "" && normCodeset != codeset {
+			mask |= xpgNormCodeset
+		}
+	}
+	pos = strings.IndexRune(locale, '_')
+	if pos != -1 {
+		territory = locale[pos:]
+		locale = locale[:pos]
+		mask |= xpgTerritory
+	}
+	language = locale
+
+	for i := mask; i >= 0; i-- {
+		if i&^mask != 0 {
+			// Only consider cases where all components
+			// are available
+			continue
+		}
+		if i&xpgNormCodeset != 0 && i&xpgCodeset != 0 {
+			// Don't include both versions of the codeset
+			continue
+		}
+		locale := language
+		if i&xpgTerritory != 0 {
+			locale += territory
+		}
+		if i&xpgCodeset != 0 {
+			locale += codeset
+		}
+		if i&xpgNormCodeset != 0 {
+			locale += normCodeset
+		}
+		if i&xpgModifier != 0 {
+			locale += modifier
+		}
+		locales = append(locales, locale)
+	}
+	return locales
+}
+
+var osGetenv = os.Getenv
+
+// localeFromEnv returns the locales set in the environment
+// This lookup is based on the logic of libintl's guess_category_value()
+func localeFromEnv() []string {
+	// libintl uses $LANGUAGE by default as a colon-separated list
+	// of locale names.
+	locale := osGetenv("LANGUAGE")
+	if len(locale) != 0 {
+		return strings.Split(locale, ":")
+	}
+
+	// It falls back to POSIX locale's LC_MESSAGES category, which
+	// is controlled by the first of $LC_ALL, $LC_MESSAGES, and
+	// $LANG.  Each of those contains a single locale.
+	for _, env := range []string{"LC_ALL", "LC_MESSAGES", "LANG"} {
+		locale = osGetenv(env)
+		if len(locale) != 0 {
+			return []string{locale}
+		}
+	}
+
+	// libintl also includes some platform specific fallbacks for
+	// Windows and MacOS that we have not implemented.
+	return nil
+}
+
+// UserLanguages returns a list of the user's preferred languages
+//
+// These are in the form of POSIX locale identifiers.
+func UserLanguages() []string {
+	var locales []string
+	seen := make(map[string]bool)
+	for _, singleLocale := range localeFromEnv() {
+		if singleLocale == "C" || singleLocale == "POSIX" {
```

Comment:
Should we support "C.UTF-8" here explicitly too?

Reply:
It doesn't look like libintl uses it as a stop locale.  It seems worth doing though.

Reply:
I've split the locale normalization out from the `UserLanguages` method, since we probably want to perform the same normalization for calls to `Translations.Locale` in the other PR where we add support for fallback catalogs.